pkg/services/ngalert/state/manager.go

```diff
@@ -289,30 +292,54 @@ func (st *Manager) ResetStateByRuleUID(ctx context.Context, rule *ngModels.Alert
 
 // ProcessEvalResults updates the current states that belong to a rule with the evaluation results.
 // if extraLabels is not empty, those labels will be added to every state. The extraLabels take precedence over rule labels and result labels
-func (st *Manager) ProcessEvalResults(ctx context.Context, evaluatedAt time.Time, alertRule *ngModels.AlertRule, results eval.Results, extraLabels data.Labels) []StateTransition {
+// This will update the states in cache/store and return the state transitions that need to be sent to the alertmanager.
+func (st *Manager) ProcessEvalResults(
+	ctx context.Context,
+	evaluatedAt time.Time,
+	alertRule *ngModels.AlertRule,
+	results eval.Results,
+	extraLabels data.Labels,
+	send Sender,
+) StateTransitions {
 	utcTick := evaluatedAt.UTC().Format(time.RFC3339Nano)
-	tracingCtx, span := st.tracer.Start(ctx, "alert rule state calculation", trace.WithAttributes(
+	ctx, span := st.tracer.Start(ctx, "alert rule state calculation", trace.WithAttributes(
 		attribute.String("rule_uid", alertRule.UID),
 		attribute.Int64("org_id", alertRule.OrgID),
 		attribute.Int64("rule_version", alertRule.Version),
 		attribute.String("tick", utcTick),
 		attribute.Int("results", len(results))))
 	defer span.End()
 
-	logger := st.log.FromContext(tracingCtx)
+	logger := st.log.FromContext(ctx)
 	logger.Debug("State manager processing evaluation results", "resultCount", len(results))
-	states := st.setNextStateForRule(tracingCtx, alertRule, results, extraLabels, logger)
+	states := st.setNextStateForRule(ctx, alertRule, results, extraLabels, logger)
+
+	staleStates := st.deleteStaleStatesFromCache(ctx, logger, evaluatedAt, alertRule)
 	span.AddEvent("results processed", trace.WithAttributes(
 		attribute.Int64("state_transitions", int64(len(states))),
+		attribute.Int64("stale_states", int64(len(staleStates))),
 	))
 
-	staleStates := st.deleteStaleStatesFromCache(ctx, logger, evaluatedAt, alertRule)
-	st.persister.Sync(tracingCtx, span, states, staleStates)
+	allChanges := StateTransitions(append(states, staleStates...))
 
-	allChanges := append(states, staleStates...)
+	// It's important that this is done *before* we sync the states to the persister. Otherwise, we will not persist
+	// the LastSentAt field to the store.
+	var statesToSend StateTransitions
+	if send != nil {
+		statesToSend = allChanges.UpdateLastSentAt(evaluatedAt)
+	}
+
+	st.persister.Sync(ctx, span, allChanges)
 	if st.historian != nil {
-		st.historian.Record(tracingCtx, history_model.NewRuleMeta(alertRule, logger), allChanges)
+		st.historian.Record(ctx, history_model.NewRuleMeta(alertRule, logger), allChanges)
 	}
+
+	// Optional callback intended for sending the states to an alertmanager.
+	// Some uses ,such as backtesting or the testing api, do not send.
+	if send != nil {
```

Comment:
I think this make logic kinda awkward: we set LastSentAt but then do not send because sender is not specified. I wonder if we should make it mandatory and those that do not want to "send" will provide Noop

Reply:
Both `UpdateLastSentAt` and `send` are gated by `send != nil`, so they will only only ever happen together.

Reply:
Oh, I totally missed it! Sorry